### PR TITLE
Darkchambers Implementation

### PR DESCRIPTION
**Pull request by Ju-li-us**
Hello,

below you’ll find our implementation! We tried to stay as faithful as possible to the ALE implementation (but there are some minor misalignments described below).

Here a rough description of the core functionality

### Items

- **Heart (`ITEM_HEART`)**: Restores player health.  
- **Poison (`ITEM_POISON`)**: Damages player health. *(we used a box scaled version of the skull sprite. not sure about that in the ALE reference game, I think it can easily be confused with a spawner there.)*  
- **Trap (`ITEM_TRAP`)**: Looks like a ladder and inflicts greater health damage.  
- **Strongbox (`ITEM_STRONGBOX`)**: Awards 100 points.  
- **Amber Chalice (`ITEM_AMBER_CHALICE`)**: Awards 500 points.  
- **Amulet (`ITEM_AMULET`)**: Awards 1000 points.  
- **Gold Chalice (`ITEM_GOLD_CHALICE`)**: Awards 3000 points.  
- **Shield (`ITEM_SHIELD`)**: Reduces damage taken.  
- **Gun (`ITEM_GUN`)**: Increases player fire rate.  
- **Bomb (`ITEM_BOMB`)**: Destroys all enemies in an area after picked up and space pressed
- **Key (`ITEM_KEY`)**: Unlocks gated areas or cage doors.  
- **Ladder Up (`ITEM_LADDER_UP`)**: Exit to the next level.  
- **Ladder Down (`ITEM_LADDER_DOWN`)**: Descend to a lower level.  
- **Cage Door (`ITEM_CAGE_DOOR`)**: Gates entry to a 3×3 cage containing a reward.  

> All items are spawned randomly and can also be dropped by spawners or enemies when defeated.  We currently spawn more items and enemies than in the ALE implementation to make testing easier  (in the ALE game it can take >1h of playtime until a bomb is found). Also shooting is currently faster to make sure the player does not die unintendedly (shooting speed can be easily lowered by a constant)

### Map

- The map consists of three hard-coded parts: **left**, **middle**, and **right**.  
- Players can traverse between parts using left/right portals, forming a circular loop (e.g., moving right from the right map wraps to the left map).  
- Each level applies a static offset to the map to create visual variation (7 levels in total so far, in the highest level ladder to moving up won't work).
- Entering a map part triggers enemy respawns within that part.
-Each map part currently has a 3×3 locked cage with a reward inside. This shows up more often than the hidden rooms in the ALE version, again to make testing easier.

We’d love to hear any suggestions or ideas you might have  — your feedback would be greatly appreciated!

Thanks so much,
Julius, Paolo, Jakob

### CODE REVIEW

**Comment by ahmetzahitak** [comment]
# Dark Chambers First Submission Review

## 1- ALE Similarity: 2.5/4
- The map layout, the sprites, and the gameplay are similar enough to the original game.
- Free space in the map sometimes feels much more than the original, it could be more like a maze. Also, the original map has many more variations in room designs.
- Health bar is visualized with not only a bar but also a number, which I find helpful.
- Enemies and items are quite similar (sometimes identical) to the original game, but dynamic movements of characters are lacking (movements of heads or footsteps).
- Picked items are just colored blocks instead of detailed sprites, which reduces the understanding.
- When two objects are too close to each other, their black backgrounds can be seen (one's background oveershadows the other). This also happens to level and health texts. (probably due to padding)
- In the original game, if an enemy is close enough to the player, it starts attacking, the player loses some health, but the enemy dies at some point. In this implementation, the enemy never dies and keeps attacking the player until the player dies.
- Enemies changes sprites (and finally die) at every hit, but in the original game, they change sprites at every 2 or 3 hits.
- The level and map changes has no animations or transitions in contrast to the original game.
- There is a bug where sometimes the right corner of the map is not rendered at all (the colored bar is missing). And when the player comes to a position (possibly where the small hole should be (where the map transition happens)), the maps changes rapidly back and forth. I could reproduce this by starting the game, go the level up ladder (I think it's always at the right bottom corner) until level 3, then go to the right corner of the map.
- Overall, the gameplay feels a bit faster and a bit harder than the original game, since enemies always chase the player directly without any random movements.

## 2- Implementation Quality : 3/4
- First of all, I didn't realize any delays or lags during gameplay, so everything seems to be running smoothly (both the rendering and the game logic).
- All of the game mechanics seems to be implemented with jax operations: jax.vmap for enemy step and collision detection, jax.lax.fori_loop and jax.lax.scan for iterations, jax.lax.cond for conditional checks.
- The render function (around 700 lines) and especially step function (around 1600 lines) is quite monolithic and long, which makes it hard to read and understand the code. And it would probably make it hard to develop some new features or modes. But there are also some nested functions inside those functions, so it helps with the readability and probably refactoring the code to make it more modular.

**Comment by M4GI3R** [comment]
# Review Darkchambers
#### From the Journey Escape Group
## ALE-Similarity

> How close does the gameplay feel to the original? Is it similar enough so both games offer the same gameplay challenge? Or are they completely different?

This is a bit tricky to evaluate, since (as you mentioned in your PR description) your current implementation is more in a testing state (like a sandbox-mode) and therefore can't be easily compared to the gameplay loop of the ALE-version yet. That's why we just add some differences we found to the ones you already listed in your description.

### Differences listed in the PR-description:

|  | JAX               |      ALE          |
|---------------|-------------------|:-----------------:|
|Enemy and Item Spawn | random |  some predefined enemy spawns <br>at the start + some hardcoded enemies &<br> items. most seem random|
|Spawn Frequency | Spawns more items and enemies than in <br>the ALE implementation to make testing easier |    way less enemies   |
|Shooting| faster -> beam shooting | one bullet shots |

### Differences we found:
|                   | JAX               |      ALE          |
|-------------------|:-----------------:|:-----------------:|
|Image Size|160, 160|210, 160|
|UI-Layout|Inside the play area| underneath the play area|
|Background Color| black | grey|
|Player Sprite | static |walking and turning animation|
|Movement Speed|faster| slower|
|Level Size|level width > screen width|level width == screen width|
|Camera Behavior| player always x,y centered| player only y centered, <br>but players x position changes

### Comments
- Adjusting the Level Size and Camera Behavior will probably bring your implementation a lot closer to ALE in terms of general gameplay feeling

### Bugs we noticed:
|                   |                |
|-------------------|:-----------------:|
| items and enemies are renderd in front of the Level lable | level_render_bug |
| the sprites of the spawner and some other enemies have black borders (instead of transparent) | Spawner_sprite_transparency |
| Player sometimes overlaps with walls after entering a new level (in particular, entering level 6 in our case). The player is then not able to move. |bug_3|
| In Level 4, the walls on the right side leave a vertical gap. If the player moves into this gap and keeps walking right, enemies and items start respawning every frame at random positions. | bug_4|


## Implementation Quality
> Is the code efficient or completely bloated? Did the environment need 50 seconds to load in even though its gameplay loop is simple? Is the code modular or obviously monolithic? Is everything jax native or were there obvious jax antipatterns like excessive sequential processing?

- JAX is well used with a lot of vmaps

- The step function (in particular handle normal) is rather verbose and less readable. Some parts could be written as helper function outside the step function (as described in section 4 "Decompose Logic into Helper Functions" of Design Guidelines) for better readability.  
  - E.g. function enemy_step, code that involves player actions (movement and bullets) into player_step, overlap/collision functions
  - Maybe the collision helper functions can be reused e.g. check_bullet_enemy_collision und check_bullet_spawner_collision; check_all_spawners_for_bullet and check_all_enemies_for_bullet
